src/services/resource.js

```diff
@@ -0,0 +1,1297 @@
+  const removedReducedResources = currentResources
+    .reduce((resources, resource) => {
+      const isRemoved = incomingResources
+        .filter((rff) => rff.genericId === resource.genericId
+          && rff.resourceId === resource.resourceId)
+        .length === 0;
+      if (isRemoved && resource.onApprovedAR !== true) {
+        const removed = resources.removed
+          ?.find((r) => r.genericId === resource.genericId);
+        if (removed) {
+          removed.resourceIds = [...removed.resourceIds, resource.resourceId];
+          return resources;
+        }
+        return {
+          removed: [
+            ...resources.removed,
+            {
+              genericId: resource.genericId,
+              resourceIds: [resource.resourceId],
+            },
+          ],
+          reduced: resources.reduced,
+        };
+      }
+
+      const matchingFromFields = incomingResources
+        .filter((rff) => rff.genericId === resource.genericId
+        && rff.resourceId === resource.resourceId);
+      const isReduced = matchingFromFields
+        .filter((mff) => resource.sourceFields
+          .filter((l) => mff.sourceFields.includes(l))
+          .length < resource.sourceFields.length)
+        .length > 0;
+      if (isReduced) {
+        const reduced = resources.reduced
+          ?.find((r) => r.genericId === resource.genericId
+            && r.resourceId === resource.resourceId);
+        const matching = matchingFromFields
+          .find((r) => r.genericId === resource.genericId
+            && r.resourceId === resource.resourceId);
+        if (reduced) {
+          reduced.sourceFields = reduced.sourceFields
+            .filter((sourceField) => matching.sourceFields.includes(sourceField));
+          return resources;
+        }
+        return {
+          removed: resources.removed,
+          reduced: [
+            ...resources.reduced,
+            {
+              genericId: resource.genericId,
+              resourceId: resource.resourceId,
+              sourceFields: resource.sourceFields
+                .filter((sourceField) => matching.sourceFields.includes(sourceField)),
+            },
+          ],
+        };
+      }
+
+      return resources;
+    }, { removed: [], reduced: [] });
+
+  // collect the intersection of the expanded and reduced datasets to generate the delta dataset.
+  const deltaFromExpanded = (newExpandedResources.expanded || [])
+    .filter((neResource) => (removedReducedResources.reduced || [])
+      .filter((rrResource) => neResource.genericId === rrResource.genericId
+        && neResource.resourceId === rrResource.resourceId)
+      .length > 0);
+  const deltaFromReduced = (removedReducedResources.reduced || [])
+    .filter((rrResource) => (newExpandedResources.expanded || [])
+      .filter((neResource) => neResource.genericId === rrResource.genericId
+        && neResource.resourceId === rrResource.resourceId)
+      .length > 0);
```

Comment:
you can use `some` instead of checking the length of the filtered resources, which should reduce the cognitive complexity of these methods. using `some` instead of `filter` means we also avoid creating a new array just to check its length.

```suggestion
const deltaFromExpanded = (newExpandedResources.expanded || []).filter((neResource) => {
  const isRemoved = (removedReducedResources.reduced || []).some((rrResource) => (
    neResource.genericId === rrResource.genericId && neResource.resourceId === rrResource.resourceId
  ));
  return isRemoved;
});

const deltaFromReduced = (removedReducedResources.reduced || []).filter((rrResource) => {
  const isAdded = (newExpandedResources.expanded || []).some((neResource) => (
    neResource.genericId === rrResource.genericId && neResource.resourceId === rrResource.resourceId
  ));
  return isAdded;
});
```